controllers/provisioning_controller.go

```diff
@@ -222,49 +219,42 @@ func (r *ProvisioningReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
 		return ctrl.Result{}, errors.Wrap(err, "failed to create Inspector password")
 	}
 
-	// If Metal3 Deployment already exists, do nothing.
-	exists, err := provisioning.GetExistingMetal3Deployment(r.KubeClient.AppsV1(), ComponentNamespace)
-	if err != nil && !apierrors.IsNotFound(err) {
-		return ctrl.Result{}, errors.Wrap(err, "failed to find existing Metal3 Deployment")
-	}
-
-	if exists {
-		r.Log.V(1).Info("metal3 deployment already exists")
-		err = r.updateCOStatus(ReasonComplete, "found existing Metal3 deployment", "")
-		if err != nil {
-			return ctrl.Result{}, fmt.Errorf("unable to put %q ClusterOperator in Available state: %v", clusterOperatorName, err)
-		}
-		return ctrl.Result{}, nil
-	}
-
-	err = r.updateCOStatus(ReasonSyncing, "", "Applying the Metal3 deployment")
+	// Update status of CO to indicate that it is progressing towards a metal3 deployment
+	err = r.updateCOStatus(ReasonSyncing, "syncing metal3 pod", "")
 	if err != nil {
-		return ctrl.Result{}, fmt.Errorf("unable to put %q ClusterOperator in Syncing state: %v", clusterOperatorName, err)
+		return ctrl.Result{}, fmt.Errorf("unable to put %q ClusterOperator in Progressing state: %v", clusterOperatorName, err)
 	}
 
-	// Proceed with creating a new Metal3 deployment
+	// Create metal3 deployment object based on current baremetal configuration
 	metal3Deployment := provisioning.NewMetal3Deployment(ComponentNamespace, &containerImages, &baremetalConfig.Spec)
-	expectedGeneration := resourcemerge.ExpectedDeploymentGeneration(metal3Deployment, r.Generations)
 
-	err = controllerutil.SetControllerReference(baremetalConfig, metal3Deployment, r.Scheme)
+	// Apply new deployment or compare against existing deployment and update
+	err = provisioning.ApplyNewOrUpdatedDeployment(r.KubeClient.AppsV1(), ComponentNamespace, metal3Deployment, r.Generations, ComponentName)
```

Comment:
Will `ApplyNewOrUpdatedDeployment` overwrite the specs of an already existing deployment? If so, I'd recommend to keep a simpler logic, ie if the deployment does not exist let's create it, otherwise just annotate it

Reply:
ApplyNewOrUpdatedDeployment() needs to do one other thing. As part of allowing changes to Provisioning CR on day 2, an existing metal3 deployment has to adjust to that. That is why I have that logic around comparing the Spec fields of existing metal3 deployment and the new metal3Deployment object just created using the latest Provisioning CR.

Reply:
My main concern is about supporting two scenarios:
1) Already existing metal3 deployment (and that's the most common one, during the upgrade)
2) Manual emergency change to metal3 deployment (ie an environment variable of one pod's container)
In such cases I don't think CBO should overwrite the existing metal3 specs with new values

Reply:
The whole point of this operator is to manage the metal3 deployment. It should definitely overwrite whatever is there, entirely, if it is different from what it thinks it should be. That's it's #1 job.

If a user needs to change something in the deployment, we need to give them an API for that. If they want to make manual changes, they should turn the cluster-baremetal-operator off and recognize that they're outside of the supported configuration for a cluster.

Reply:
Ok, that's fine, I didn't want to remove the ownership to CBO or end up in a scenario without it. I was looking to how to support a scenario like https://github.com/metal3-io/metal3-docs/blob/master/design/baremetal-operator/limit-hosts-provisioning.md#implementation-detailsnotesconstraints, where a configuration variable is present but not exposed in the complete chain (ie from the installer up). Could make sense to have some custom fields in the Provisioning CR for that?

Reply:
I think the way to handle that would be to pass those environment variables we want to make configurable from an optional ConfigMap that the user can supply.